From 21e1fe692ef2d453244d8396b8ff4286e7f98c76 Mon Sep 17 00:00:00 2001
From: Madhurkiran Harikrishnan <madhurkiran.harikrishnan@xilinx.com>
Date: Fri, 26 Apr 2019 17:24:16 -0700
Subject: [PATCH] gator: Make the module compatilble with kernel >=4.19

Refer kernel patch 9c0be3f6, which fixes tracepoint array element size
mismatch. This patch leverages the same.

Signed-off-by: Madhurkiran Harikrishnan <madhurkiranharikrishnan@gmail.com>
---
Upstream-Status: Pending

 driver/gator_main.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/driver/gator_main.c b/driver/gator_main.c
index 67de98e..66ef75d 100644
--- a/driver/gator_main.c
+++ b/driver/gator_main.c
@@ -1399,21 +1399,34 @@ GATOR_TRACEPOINTS;
 int gator_new_tracepoint_module(struct notifier_block * nb, unsigned long action, void * data)
 {
     struct tp_module * tp_mod = (struct tp_module *) data;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0)
+    tracepoint_ptr_t * begin = tp_mod->mod->tracepoints_ptrs;
+    tracepoint_ptr_t * end = tp_mod->mod->tracepoints_ptrs + tp_mod->mod->num_tracepoints;
+#else
     struct tracepoint * const * begin = tp_mod->mod->tracepoints_ptrs;
     struct tracepoint * const * end = tp_mod->mod->tracepoints_ptrs + tp_mod->mod->num_tracepoints;
+#endif
 
     pr_debug("gator: new tracepoint module registered %s\n", tp_mod->mod->name);
 
     if (action == MODULE_STATE_COMING)
     {
         for (; begin != end; ++begin) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0)
+            gator_save_tracepoint(tracepoint_ptr_deref(begin), NULL);
+#else
             gator_save_tracepoint(*begin, NULL);
+#endif
         }
     }
     else if (action == MODULE_STATE_GOING)
     {
         for (; begin != end; ++begin) {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 19, 0)
+            gator_unsave_tracepoint(tracepoint_ptr_deref(begin), NULL);
+#else
             gator_unsave_tracepoint(*begin, NULL);
+#endif
         }
     }
     else
